modelservice/worker.py

```diff
@@ -72,12 +78,118 @@ def register(self):
+            if normalize_inputs:
+                scaled = MinMaxScaler(feature_range=(0, 1)).fit_transform(all_df)
+                all_df = pd.DataFrame(scaled, columns=all_df.columns)
+
+            features = all_df[feature_fields]
+            labels = all_df[label_field]
+
+            # Create Sequential model
+            model = tf.keras.Sequential()
+
+            # Add input layer
+            model.add(tf.keras.Input(shape=(len(request.feature_fields))))
+
+            # Add hidden layers
+            for hidden_layer in hyper_parameters.hidden_layers:
+                name: str = hidden_layer.name
+                units: int = hidden_layer.units
+                activation = "relu"
+                model.add(tf.keras.layers.Dense(units=units, activation=activation, name=name))
+
+            # Add output layer
+            output_layer: OutputLayer = hyper_parameters.output_layer
+            name: str = output_layer.name
+            activation = "relu"
+            model.add(tf.keras.layers.Dense(units=1, activation=activation, name=name))
+
+            # Compile the model and print its summary
+            model.compile(loss=loss, optimizer=optimizer)
+            model.summary()
+
+            # Fit the model to the data
+            history = model.fit(features, labels, epochs=epochs, validation_split=test_split)
+            hist = pd.DataFrame(history.history)
+            hist["epoch"] = history.epoch
+            info(hist)
+
+            last_row = hist.loc[hist["epoch"] == epochs - 1].values[0]
+            training_loss = last_row[0]
+            validation_loss = last_row[1]
+            info(f"Training loss: {training_loss}, validation loss: {validation_loss}")
+            gis_join_timer.stop()
+
+            metric: EvaluationMetric = EvaluationMetric(
+                training_loss=training_loss,
+                validation_loss=validation_loss,
+                duration_sec=gis_join_timer.elapsed,
+                error_occurred=False,
+                error_message="",
+                gis_join_metadata=GisJoinMetadata(
+                    gis_join=gis_join,
+                    count=len_df
+                )
+            )
+            evaluation_metrics.append(metric)
+            model_path: str = f"{models_dir}/{gis_join}.tf"
+            model.save(model_path)
+            info(f"Saved model {model_path}")
+            count += 1
+            gis_join_timer.reset()
+
+        info(f"Finished training {count}/{len(request.gis_joins)} models. Returning results...")
+        worker_timer.stop()
+        return WorkerBuildModelsResponse(
             id=request.id,
             hostname=self.hostname,
-            duration_sec=0.0,  # TODO: Capture job profile
-            error_occurred=True,
-            error_msg="Building models currently unimplemented"
+            duration_sec=worker_timer.elapsed,
+            error_occurred=False,
```

Comment:
Not sure that we need to, but do we want to add some logic for handling errors?

Reply:
Maybe if we were putting this into production ;)